<center>
<img src="../../img/ods_stickers.jpg" />
    
## [mlcourse.ai](https://mlcourse.ai) – Open Machine Learning Course 

Author: [Yury Kashnitsky](https://yorko.github.io). This material is subject to the terms and conditions of the [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/) license. Free use is permitted for any non-commercial purpose.

# <center> Topic 1. Exploratory data analysis with Pandas
## <center>Practice. Analyzing "Titanic" passengers

**Fill in the missing code ("You code here") and choose answers in a [web-form](https://docs.google.com/forms/d/16EfhpDGPrREry0gfDQdRPjoiQX9IumaL2mPR0rcj19k/edit).**

In [1]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
# Graphics in SVG format are more sharp and legible
%config InlineBackend.figure_format = 'svg'
pd.set_option("display.precision", 2)

In [162]:
np.set_printoptions(precision=2)

**Read data into a Pandas DataFrame**

In [42]:
# data = pd.read_csv('data/titanic.csv', sep="\t")
data = pd.read_csv('data/titanic_2.csv')

In [43]:
data

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket   Fare Cabin Embarked  
0        0         A/5 21171   7.25   NaN        S  
1        0          PC 17599  71.28   C85        C  
2        0  STON/O2. 3101282   7.92   NaN        S  
3        0            113803  53.10  C123        S  
4        0            373450   8.05   NaN        S  
..     ...               ...    ...   ...      ...  
886      0            211536  13.00   NaN        S  
887      0            112053  30.00   B42        S  
888      2        W./C. 6607  23.45   NaN        S  
889      0            111369  30.00  C148        C  
890      0            370376   7.75   NaN        Q  

[891 rows x 12 columns]

**First 5 rows**

In [44]:
data.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket   Fare Cabin Embarked  
0      0         A/5 21171   7.25   NaN        S  
1      0          PC 17599  71.28   C85        C  
2      0  STON/O2. 3101282   7.92   NaN        S  
3      0            113803  53.10  C123        S  
4      0            373450   8.05   NaN        S

In [45]:
data.describe()

PassengerId  Survived  Pclass     Age   SibSp   Parch    Fare
count       891.00    891.00  891.00  714.00  891.00  891.00  891.00
mean        446.00      0.38    2.31   29.70    0.52    0.38   32.20
std         257.35      0.49    0.84   14.53    1.10    0.81   49.69
min           1.00      0.00    1.00    0.42    0.00    0.00    0.00
25%         223.50      0.00    2.00   20.12    0.00    0.00    7.91
50%         446.00      0.00    3.00   28.00    0.00    0.00   14.45
75%         668.50      1.00    3.00   38.00    1.00    0.00   31.00
max         891.00      1.00    3.00   80.00    8.00    6.00  512.33

**Let's select those passengers who embarked in Cherbourg (Embarked=C) and paid > 200 pounds for their ticker (fare > 200).**

Make sure you understand how actually this construction works.

In [59]:
data.sort_values(by="Fare", ascending =False).head()

PassengerId  Survived  Pclass                                Name  \
258          259         1       1                    Ward, Miss. Anna   
737          738         1       1              Lesurer, Mr. Gustave J   
679          680         1       1  Cardeza, Mr. Thomas Drake Martinez   
88            89         1       1          Fortune, Miss. Mabel Helen   
27            28         0       1      Fortune, Mr. Charles Alexander   

        Sex   Age  SibSp  Parch    Ticket    Fare        Cabin Embarked  \
258  female  35.0      0      0  PC 17755  512.33          NaN        C   
737    male  35.0      0      0  PC 17755  512.33         B101        C   
679    male  36.0      0      1  PC 17755  512.33  B51 B53 B55        C   
88   female  23.0      3      2     19950  263.00  C23 C25 C27        S   
27     male  19.0      3      2     19950  263.00  C23 C25 C27        S   

     Age_category  
258           2.0  
737           2.0  
679           2.0  
88            1.0  
27            1.0

In [163]:
data[(data['Embarked'] == 'C') & (data.Fare > 200)].head()

PassengerId  Survived  Pclass  \
118          119         0       1   
258          259         1       1   
299          300         1       1   
311          312         1       1   
377          378         0       1   

                                                Name     Sex   Age  SibSp  \
118                         Baxter, Mr. Quigg Edmond    male  24.0      0   
258                                 Ward, Miss. Anna  female  35.0      0   
299  Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0      0   
311                       Ryerson, Miss. Emily Borie  female  18.0      2   
377                        Widener, Mr. Harry Elkins    male  27.0      0   

     Parch    Ticket    Fare            Cabin Embarked  Age_category  
118      1  PC 17558  247.52          B58 B60        C           1.0  
258      0  PC 17755  512.33              NaN        C           2.0  
299      1  PC 17558  247.52          B58 B60        C           2.0  
311      2  PC 17608  262.38  B57 B59 B63 B66        C           1.0  
377      2    113503  211.50              C82        C           1.0

**We can sort these people by Fare in descending order.**

In [49]:
data[(data['Embarked'] == 'C') & 
     (data['Fare'] > 200)].sort_values(by='Fare',
                               ascending=False).head()

PassengerId  Survived  Pclass                                   Name  \
258          259         1       1                       Ward, Miss. Anna   
679          680         1       1     Cardeza, Mr. Thomas Drake Martinez   
737          738         1       1                 Lesurer, Mr. Gustave J   
311          312         1       1             Ryerson, Miss. Emily Borie   
742          743         1       1  Ryerson, Miss. Susan Parker "Suzette"   

        Sex   Age  SibSp  Parch    Ticket    Fare            Cabin Embarked  
258  female  35.0      0      0  PC 17755  512.33              NaN        C  
679    male  36.0      0      1  PC 17755  512.33      B51 B53 B55        C  
737    male  35.0      0      0  PC 17755  512.33             B101        C  
311  female  18.0      2      2  PC 17608  262.38  B57 B59 B63 B66        C  
742  female  21.0      2      2  PC 17608  262.38  B57 B59 B63 B66        C

**Let's create a new feature.**

In [37]:
def age_category(age):
    '''
    < 30 -> 1
    >= 30, <55 -> 2
    >= 55 -> 3
    '''
    if age < 30:
        return 1
    elif age < 55:
        return 2
    elif age >= 55:
        return 3

In [54]:
age_categories = [age_category(age) for age in data.Age]
data['Age_category'] = age_categories

**Another way is to do it with `apply`.**

In [55]:
data['Age_category'] = data['Age'].apply(age_category)

**1. How many men/women were there onboard?**
- 412 men and 479 women
- 314 men and 577 women
- 479 men and 412 women
- 577 men and 314 women

In [108]:
data.groupby('Sex').count()

PassengerId  Survived  Pclass  Name  Age  SibSp  Parch  Ticket  Fare  \
Sex                                                                            
female          314       314     314   314  261    314    314     314   314   
male            577       577     577   577  453    577    577     577   577   

        Cabin  Embarked  Age_category  
Sex                                    
female     97       312           261  
male      107       577           453

In [165]:

data.Embarked.describe()

count     889
unique      3
top         S
freq      644
Name: Embarked, dtype: object

**2. Print the distribution of the `Pclass` feature. Then the same, but for men and women separately. How many men from second class were there onboard?**
- 104
- 108
- 112
- 125

In [166]:
data.loc[data.Embarked.notna()].groupby(['Pclass','Sex']).count()

PassengerId  Survived  Name  Age  SibSp  Parch  Ticket  Fare  \
Pclass Sex                                                                    
1      female           92        92    92   83     92     92      92    92   
       male            122       122   122  101    122    122     122   122   
2      female           76        76    76   74     76     76      76    76   
       male            108       108   108   99    108    108     108   108   
3      female          144       144   144  102    144    144     144   144   
       male            347       347   347  253    347    347     347   347   

               Cabin  Embarked  Age_category  
Pclass Sex                                    
1      female     79        92            83  
       male       95       122           101  
2      female     10        76            74  
       male        6       108            99  
3      female      6       144           102  
       male        6       347           253

**3. What are median and standard deviation of `Fare`?. Round to two decimals.**
- median is  14.45, standard deviation is 49.69
- median is 15.1, standard deviation is 12.15
- median is 13.15, standard deviation is 35.3
- median is  17.43, standard deviation is 39.1

In [172]:
np.round(data.Fare.median(),decimals=2)

14.45

**4. Is that true that the mean age of survived people is higher than that of passengers who eventually died?**
- Yes
- No


In [143]:
if(data.loc[(data.Sex == 'male') & (data.Survived == 1)].Age.mean() > data.loc[data.Survived == 0].Age.mean()):
    print('Yes') 
else: 
    print('No') 

No


**5. Is that true that passengers younger than 30 y.o. survived more frequently than those older than 60 y.o.? What are shares of survived people among young and old people?**
- 22.7% among young and 40.6% among old
- 40.6% among young and 22.7% among old
- 35.3% among young and 27.4% among old
- 27.4% among young and  35.3% among old

In [148]:
data.loc[data.Age_category==1].Survived.mean() > data.loc[data.Age_category==3].Survived.mean()

True

In [178]:
data.groupby('Age_category').Survived.mean()*100

Age_category
1.0    40.62
2.0    42.01
3.0    30.95
Name: Survived, dtype: float64

**6. Is that true that women survived more frequently than men? What are shares of survived people among men and women?**
- 30.2% among men and 46.2% among women
- 35.7% among men and 74.2% among women
- 21.1% among men and 46.2% among women
- 18.9% among men and 74.2% among women

In [158]:
data.loc[data.Sex=='female'].Survived.mean() > data.loc[data.Sex=='male'].Survived.mean()

True

In [197]:
data.groupby('Sex').Survived.mean()*100

Sex
female    74.20
male      18.89
Name: Survived, dtype: float64

**7. What's the most popular first name among male passengers?**
- Charles
- Thomas
- William
- John

In [215]:
def extractFirstName(name) :
    return name[name.find(',')+1:].replace('Mr.','').replace('Mrs.','').replace('Miss.','').strip().split(' ')[0]

In [257]:
first_name = [extractFirstName(name) for name in data.Name]
data['First_name'] = first_name

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket   Fare Cabin Embarked  Age_category First_name  
0      0         A/5 21171   7.25   NaN        S           1.0       Owen  
1      0          PC 17599  71.28   C85        C           2.0       John  
2      0  STON/O2. 3101282   7.92   NaN        S           1.0      Laina  
3      0            113803  53.10  C123        S           2.0    Jacques  
4      0            373450   8.05   NaN        S           2.0    William

In [241]:
data.First_name.value_counts()

William    43
Master.    40
John       27
Thomas     17
Charles    15
           ..
Fared       1
Marin       1
Bengt       1
Julie       1
Phyllis     1
Name: First_name, Length: 426, dtype: int64

**8. How is average age for men/women dependent on `Pclass`? Choose all correct statements:**
- On average, men of 1 class are older than 40
- On average, women of 1 class are older than 40
- Men of all classes are on average older than women of the same class
- On average, passengers of the first class are older than those of the 2nd class who are older than passengers of the 3rd class

In [246]:
print(data.loc[ (data.Sex == 'male') & (data.Pclass == 1) ].Age.mean() > 40)
print(data.loc[ (data.Sex == 'female') & (data.Pclass == 1) ].Age.mean() > 40)

True

In [297]:
vista = data.groupby(['Pclass','Sex']).agg({"Age":"mean"})


In [300]:
vista[vista.Sex == 'female'

SyntaxError: unexpected EOF while parsing (<ipython-input-300-1e46dac4d248>, line 1)

In [ ]:
data.agg("")

## Useful resources
* The same notebook as an interactive web-based [Kaggle Kernel](https://www.kaggle.com/kashnitsky/topic-1-practice-analyzing-titanic-passengers) with a [solution](https://www.kaggle.com/kashnitsky/topic-1-practice-solution)
* Topic 1 "Exploratory Data Analysis with Pandas" as a [Kaggle Kernel](https://www.kaggle.com/kashnitsky/topic-1-exploratory-data-analysis-with-pandas)
* Main course [site](https://mlcourse.ai), [course repo](https://github.com/Yorko/mlcourse.ai), and YouTube [channel](https://www.youtube.com/watch?v=QKTuw4PNOsU&list=PLVlY_7IJCMJeRfZ68eVfEcu-UcN9BbwiX)